packages/eslint-plugin/src/rules/no-unnecessary-type-constraint.ts

```diff
@@ -0,0 +1,116 @@
+import { TSESTree } from '@typescript-eslint/experimental-utils';
+import * as semver from 'semver';
+import * as ts from 'typescript';
+import * as util from '../util';
+
+type MakeRequired<Base, Key extends keyof Base> = Omit<Base, Key> &
+  Required<Pick<Base, Key>>;
+
+type TypeParameterWithConstraint = MakeRequired<
+  TSESTree.TSTypeParameter,
+  'constraint'
+>;
+
+type KeywordFilter = (type: ts.Type) => boolean;
+
+const is3dot5 = semver.satisfies(
+  ts.version,
+  `>= 3.5.0 || >= 3.5.1-rc || >= 3.5.0-beta`,
+  {
+    includePrerelease: true,
+  },
+);
+
+const is3dot9 =
+  is3dot5 &&
+  semver.satisfies(ts.version, `>= 3.9.0 || >= 3.9.1-rc || >= 3.9.0-beta`, {
+    includePrerelease: true,
+  });
+
+export default util.createRule({
+  name: 'no-unnecessary-type-constraint',
+  meta: {
+    docs: {
+      category: 'Best Practices',
+      description: 'Disallows unnecessary constraints on generic types',
+      recommended: false,
+      requiresTypeChecking: true,
+      suggestion: true,
+    },
+    fixable: 'code',
+    messages: {
+      unnecessaryConstraint:
+        'Constraining a generic type to {{constraint}} does nothing and is unnecessary.',
+    },
+    schema: [],
+    type: 'suggestion',
+  },
+  defaultOptions: [],
+  create(context) {
+    const parserServices = util.getParserServices(context);
+    const checker = parserServices.program.getTypeChecker();
+
+    const keywordFilters: [KeywordFilter, string][] = [];
+    if (is3dot5) {
+      keywordFilters.push([util.isTypeUnknownType, 'unknown']);
+
+      if (is3dot9) {
+        keywordFilters.push([util.isTypeAnyType, 'any']);
+      }
+    }
+
+    if (!keywordFilters.length) {
+      return {};
+    }
+
+    const inJsx = context.getFilename().toLowerCase().endsWith('tsx');
+
+    const report = (
+      node: TypeParameterWithConstraint,
+      constraint: string,
+      inArrowFunction: boolean,
+    ): void => {
+      context.report({
+        data: { constraint },
+        fix(fixer) {
+          return fixer.replaceTextRange(
+            [node.name.range[1], node.constraint.range[1]],
+            inArrowFunction && inJsx ? ',' : '',
+          );
+        },
+        messageId: 'unnecessaryConstraint',
+        node,
+      });
+    };
+
+    const checkNode = (
+      node: TypeParameterWithConstraint,
+      inArrowFunction: boolean,
+    ): void => {
+      const constraint = parserServices.esTreeNodeToTSNodeMap.get(
+        node.constraint,
+      );
+      const constraintType = checker.getTypeAtLocation(constraint);
```

Comment:
do we need to do this with type info?
Could we get away with just using the keywords?

Type info restricts the usage of the rule because not everyone uses type info.

Some users also purposely use aliases for `any`/`unknown` as a "todo", so using type info would prevent that usecase.
On the flip side it would also _catch_ that use-case, but I think it'd be better to start without and add it later if requested

Reply:
For common cases we could, but cases with unions become a little difficult:

```
function data<T extends any | number>() {}
function data<T extends unknown | number>() {}
```

Is there established precedent anywhere for splitting out union literals like this, factoring in `|`s and `&`s?

Reply:
They're not too difficult - just require some AST traversal!

> Is there established precedent anywhere for splitting out union literals like this

I don't think so. Rules that do checks on type names usually just check the name and don't care about the aggregate result.

This should be simple enough though.

- `any` `|` or `&` anything is  `any`
- `unknown` `|` anything is `unknown`
- `unknown` `&` anything is the thing

Though I would question - is this a pattern people use? Have you seen this before?
I would consider maybe just ignoring these cases with a TODO comment for now (if you're feeling lazy)